vms/platformvm/state/state.go

```diff
@@ -944,74 +977,172 @@ func (s *state) ValidatorSet(subnetID ids.ID, vdrs validators.Set) error {
+			if err != nil {
+				return err
+			}
+
+			if err := applyWeightDiff(validators, nodeID, &weightDiff); err != nil {
+				return err
+			}
+		}
 	}
 
-	s.validatorWeightDiffsCache.Put(prefixStr, weightDiffs)
-	return weightDiffs, diffIter.Error()
+	return nil
 }
 
-func (s *state) GetValidatorPublicKeyDiffs(height uint64) (map[ids.NodeID]*bls.PublicKey, error) {
-	if publicKeyDiffs, ok := s.validatorPublicKeyDiffsCache.Get(height); ok {
-		return publicKeyDiffs, nil
+func applyWeightDiff(
+	vdrs map[ids.NodeID]*validators.GetValidatorOutput,
+	nodeID ids.NodeID,
+	weightDiff *ValidatorWeightDiff,
+) error {
+	vdr, ok := vdrs[nodeID]
+	if !ok {
+		// This node isn't in the current validator set.
+		vdr = &validators.GetValidatorOutput{
+			NodeID: nodeID,
+		}
+		vdrs[nodeID] = vdr
 	}
 
-	heightBytes := database.PackUInt64(height)
-	rawDiffDB := prefixdb.New(heightBytes, s.validatorPublicKeyDiffsDB)
-	diffDB := linkeddb.NewDefault(rawDiffDB)
-	diffIter := diffDB.NewIterator()
+	// The weight of this node changed at this block.
+	var err error
+	if weightDiff.Decrease {
+		// The validator's weight was decreased at this block, so in the
+		// prior block it was higher.
+		vdr.Weight, err = math.Add64(vdr.Weight, weightDiff.Amount)
+	} else {
+		// The validator's weight was increased at this block, so in the
+		// prior block it was lower.
+		vdr.Weight, err = math.Sub(vdr.Weight, weightDiff.Amount)
+	}
+	if err != nil {
+		return err
+	}
+
+	if vdr.Weight == 0 {
+		// The validator's weight was 0 before this block so they weren't in the
+		// validator set.
+		delete(vdrs, nodeID)
+	}
+	return nil
+}
+
+func (s *state) ApplyValidatorPublicKeyDiffs(
+	ctx context.Context,
+	validators map[ids.NodeID]*validators.GetValidatorOutput,
+	startHeight uint64,
+	endHeight uint64,
+) error {
+	diffIter := s.flatValidatorPublicKeyDiffsDB.NewIteratorWithStartAndPrefix(
+		getStartDiffKey(constants.PrimaryNetworkID, startHeight),
+		constants.PrimaryNetworkID[:],
+	)
 	defer diffIter.Release()
 
-	pkDiffs := make(map[ids.NodeID]*bls.PublicKey)
 	for diffIter.Next() {
-		nodeID, err := ids.ToNodeID(diffIter.Key())
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+
+		_, parsedHeight, nodeID, err := parseDiffKey(diffIter.Key())
 		if err != nil {
-			return nil, err
+			return err
+		}
+		// If the parsedHeight is less than our target endHeight, then we have
+		// fully processed the diffs from startHeight through endHeight.
+		if parsedHeight < endHeight {
+			break
+		}
+
+		vdr, ok := validators[nodeID]
+		if !ok {
+			continue
 		}
 
 		pkBytes := diffIter.Value()
-		pk, err := bls.PublicKeyFromBytes(pkBytes)
-		if err != nil {
-			return nil, err
+		if len(pkBytes) == 0 {
+			vdr.PublicKey = nil
+			continue
 		}
-		pkDiffs[nodeID] = pk
-	}
 
-	s.validatorPublicKeyDiffsCache.Put(height, pkDiffs)
-	return pkDiffs, diffIter.Error()
+		vdr.PublicKey = new(bls.PublicKey).Deserialize(pkBytes)
+	}
+	return diffIter.Error()
```

Comment:
Note: this does not fallback to the linkeddb index because the linkeddb index does not contain entries for when to remove the public key.

This means that this PR (intentionally) does not guarantee that public keys will be returned correctly for heights prior to running the new index.

This is intended, and not an issue because there are currently no production systems relying on the public key results. The proposervm does get the validator set, but does not use the public keys.

Reply:
^ Perhaps we should add this as a comment?